packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java

```
@@ -856,7 +856,7 @@ public String stopVideoRecording() {
       closeRenderer();
       captureSession.abortCaptures();
       mediaRecorder.stop();
```

It doesn't look like `MediaRecorder.stop` throws a `RuntimeException`: https://developer.android.com/reference/android/media/MediaRecorder#stop(). So, it must be another call. I think we should identify which call it is and use a separate try/catch to catch it since this catch as-is is ambiguous.

---

Thank you for your review!

According to those existing issues: 
https://github.com/flutter/flutter/issues/121459
https://github.com/flutter/flutter/issues/79306
https://github.com/flutter/flutter/issues/93686

It looks like somehow it does throws a `RuntimeException` when you calling `MediaRecorder.stop` and it does on a unit test I quote https://github.com/flutter/flutter/issues/93686#issuecomment-1852889118

with my physical phone, brand: OPPO Reno 9 PHM110,  I can reproduce this error:`Failed to handle method call (Ask Gemini)
                                         java.lang.RuntimeException: stop failed.
                                         	at android.media.MediaRecorder.stop(Native Method)
                                         	at io.flutter.plugins.camera.Camera.stopVideoRecording(Camera.java:810)
                                         	at io.flutter.plugins.camera.MethodCallHandlerImpl.onMethodCall(MethodCallHandlerImpl.java:129)
                                         	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:258)
                                         	at io.flutter.embedding.engine.dart.DartMessenger.invokeHandler(DartMessenger.java:295)
                                         	at io.flutter.embedding.engine.dart.DartMessenger.lambda$dispatchMessageToQueue$0$io-flutter-embedding-engine-dart-DartMessenger(DartMessenger.java:322)
                                         	at io.flutter.embedding.engine.dart.DartMessenger$$ExternalSyntheticLambda0.run(Unknown Source:12)
                                         	at android.os.Handler.handleCallback(Handler.java:958)
                                         	at android.os.Handler.dispatchMessage(Handler.java:99)
                                         	at android.os.Looper.loopOnce(Looper.java:257)
                                         	at android.os.Looper.loop(Looper.java:368)
                                         	at android.app.ActivityThread.main(ActivityThread.java:8839)
                                         	at java.lang.reflect.Method.invoke(Native Method)
                                         	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:572)
                                         	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1049)`
 
So I believe that this PR is a solution for this issue

---

Ah I see. I apologize; I missed the note in the docs about RuntimeException being thrown.

According to that note, sounds like we need to ensure proper cleanup is done if a RuntimeException is thrown, since the video would not be valid. I'm concerned about burying the exception versus, for example, forwarding a `FlutterError` to the developer so they can handle anything that relies on a valid video.

Let me know what you think, and @stuartmorgan if you have thoughts on this let me know.

---

Yes, this should be reported back to the Dart code somehow so that the Dart code can throw a `CameraException`.

If this is the only error case for stopping, I would suggest changing the Pigeon `stopVideoRecording()` method to return a nullable string and document that a null return means that it failed to produce video data, then have the Dart code convert that to a `CameraException`. (If there are other failures, we could use a return class instead, containing a `String?` for the file path, and an enum and/or error message for different errors.)

---

> Yes, this should be reported back to the Dart code somehow so that the Dart code can throw a `CameraException`.
> 
> If this is the only error case for stopping, I would suggest changing the Pigeon `stopVideoRecording()` method to return a nullable string and document that a null return means that it failed to produce video data, then have the Dart code convert that to a `CameraException`. (If there are other failures, we could use a return class instead, containing a `String?` for the file path, and an enum and/or error message for different errors.)

Hi Cam & stuart , Thanks for your enlightenment.
  I agree that we need to put the error to the dart code side and let the developer deal with it.

 `mediaRecorder.stop()` would throws two type of Exception:
 `IllegalStateException` if it is called before start()
  `RuntimeException` if no valid audio/ video data has been received when stop() is called. This happens if stop() is called immediately after start()
  the former code only catch and ignored the `IllegalStateException`, do you think we need to wrap these errors into a class and throw it up to dart, or we still ignore the `IllegalStateException` and return a nullable string as we handle the `RuntimeException`

---

Known cases of exceptions [should not be surfaced as raw `PlatformException`s](https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#platform-exception-handling), so both should be caught, and translated in Dart (`StateError`, in the case of the `IllegalStateException` you described).

---

> Known cases of exceptions [should not be surfaced as raw `PlatformException`s](https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#platform-exception-handling), so both should be caught, and translated in Dart (`StateError`, in the case of the `IllegalStateException` you described).

As far as I know, In Dart `stopVideoRecording()` the former code does not throw any error , So I think what we are doing would make a breaking change for developers

---

Sorry, I missed the part of your comment where you said that `IllegalStateException` is currently being caught and silently ignored.

So for the purposes of this PR, that behavior should stay as-is, and only the runtime exception should get new handling.